From e313901fc3919750056121e801c7435f0c66e420 Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Thu, 20 Sep 2018 16:46:12 +0800
Subject: [PATCH] DRVLIB_SW-930 - Change GPHY_CDB_PDI write to non-mask write

Reading this register somehow causes intermittent issues
when reading CGU registers (they become invalid until we
read back the GPHY_CDB_PDI register). Hence we replace it
with non-mask write (we do not rely on previous value anyway).
---
 drivers/net/ethernet/lantiq/xrx500_phy_fw.c | 25 ++++++-------------------
 1 file changed, 6 insertions(+), 19 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
index 510e18479c26..755cc8d9700d 100644
--- a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
+++ b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
@@ -83,6 +83,7 @@ static u32 xrx500_gphy[] = {
 #define PRX300_GPHY_CDB_PDI_PLL_CFG2 0x8
 #define PRX300_GPHY_CDB_PDI_PLL_MISC 0xc
 #define PRX300_PLL_FBDIV 0x145
+#define PRX300_PLL_LOCK_RST 0xb
 #define PRX300_PLL_REFDIV 0x4
 #define PRX300_GPHY_FORCE_LATCH 1
 #define PRX300_GPHY_CLEAR_STICKY 1
@@ -91,23 +92,11 @@ static u32 xrx500_gphy[] = {
 #define PRX300_LAN_MUX_MASK 0x2
 #define PRX300_LAN_MUX_GPHY 0x0
 
-static u32 gsw_reg_r32(void __iomem *base, u32 reg_off)
-{
-	return __raw_readl(base + reg_off);
-}
-
 static void gsw_reg_w32(void __iomem *base, u32 val, u32 reg_off)
 {
 	__raw_writel(val, base + reg_off);
 }
 
-static void gsw_reg_w32_mask(void __iomem *base, u32 clear, u32 val,
-			     u32 reg_off)
-{
-	gsw_reg_w32(base, val | (gsw_reg_r32(base, reg_off) & (~clear)),
-		    reg_off);
-}
-
 /* xrx500 specific boot sequence */
 static int xrx500_gphy_boot(struct xway_gphy_data *priv)
 {
@@ -194,15 +183,13 @@ static int prx300_gphy_boot(struct xway_gphy_data *priv)
 	reset_control_deassert(rst->gphy_pwr_down);
 
 	/* Set divider and misc config */
-	gsw_reg_w32_mask(priv->base, 0xFFF0, (PRX300_PLL_FBDIV << 4),
-			 PRX300_GPHY_CDB_PDI_PLL_CFG0);
+	gsw_reg_w32(priv->base, (PRX300_PLL_FBDIV << 4) | PRX300_PLL_LOCK_RST,
+		    PRX300_GPHY_CDB_PDI_PLL_CFG0);
 	gsw_reg_w32(priv->base, (PRX300_PLL_REFDIV << 8),
 		    PRX300_GPHY_CDB_PDI_PLL_CFG2);
-	gsw_reg_w32_mask(priv->base, (PRX300_GPHY_FORCE_LATCH << 13) |
-			 (PRX300_GPHY_CLEAR_STICKY << 14),
-			 (PRX300_GPHY_FORCE_LATCH << 13) |
-			 (PRX300_GPHY_CLEAR_STICKY << 14),
-			 PRX300_GPHY_CDB_PDI_PLL_MISC);
+	gsw_reg_w32(priv->base, (PRX300_GPHY_FORCE_LATCH << 13) |
+		    (PRX300_GPHY_CLEAR_STICKY << 14),
+		    PRX300_GPHY_CDB_PDI_PLL_MISC);
 
 	/* delay to wait until firmware boots up */
 	msleep(100);
